presto-spark-base/src/main/java/com/facebook/presto/spark/execution/nativeprocess/HttpNativeExecutionTaskResultFetcher.java

```diff
@@ -136,101 +145,90 @@ public Optional<SerializedPage> pollPage()
-
-        private int transportErrorRetries;
-        private long token;
-
-        public HttpNativeExecutionTaskResultFetcherRunner(
-                PrestoSparkHttpTaskClient client,
-                CompletableFuture<Void> future,
-                LinkedBlockingDeque<SerializedPage> pageBuffer,
-                AtomicLong bufferMemoryBytes,
-                Object taskFinishedOrHasResult)
-        {
-            this.client = requireNonNull(client, "client is null");
-            this.future = requireNonNull(future, "future is null");
-            this.pageBuffer = requireNonNull(pageBuffer, "pageBuffer is null");
-            this.bufferMemoryBytes = requireNonNull(
-                    bufferMemoryBytes,
-                    "bufferMemoryBytes is null");
-            this.taskFinishedOrHasResult = requireNonNull(taskFinishedOrHasResult, "taskFinishedOrHasResult is null");
+        if (bufferMemoryBytes.longValue() >= MAX_BUFFER_SIZE.toBytes()) {
+            return;
         }
 
-        @Override
-        public void run()
-        {
-            try {
-                if (bufferMemoryBytes.longValue() >= MAX_BUFFER_SIZE.toBytes()) {
-                    return;
-                }
-                PageBufferClient.PagesResponse pagesResponse = client.getResults(token, MAX_RESPONSE_SIZE).get();
-
-                List<SerializedPage> pages = pagesResponse.getPages();
-                long bytes = 0;
-                long positionCount = 0;
-                for (SerializedPage page : pages) {
-                    if (!isChecksumValid(page)) {
-                        throw new PrestoException(
-                                SERIALIZED_PAGE_CHECKSUM_ERROR,
-                                format("Received corrupted serialized page from host %s",
-                                        HostAddress.fromUri(client.getLocation())));
-                    }
-                    bytes += page.getSizeInBytes();
-                    positionCount += page.getPositionCount();
-                }
-                log.info("Received %s rows in %s pages from %s", positionCount, pages.size(), client.getTaskUri());
-                pageBuffer.addAll(pages);
-                bufferMemoryBytes.addAndGet(bytes);
-                long nextToken = pagesResponse.getNextToken();
-                if (pages.size() > 0) {
-                    client.acknowledgeResultsAsync(nextToken);
-                }
-                this.token = nextToken;
-                if (pagesResponse.isClientComplete()) {
-                    client.abortResults();
-                    future.complete(null);
-                }
-                if (!pages.isEmpty()) {
-                    synchronized (taskFinishedOrHasResult) {
-                        taskFinishedOrHasResult.notifyAll();
+        ListenableFuture<PageBufferClient.PagesResponse> resultsFuture = workerClient.getResults(token, MAX_RESPONSE_SIZE);
+        Futures.addCallback(
```

Comment:
Using `Futures.addCallback` would introduce an async callback inside an already async scheduled thread. This will result in the `scheduleAtFixedRate` function call believing that the work (http call) is complete, when its not. And it will schedule multiple getResults calls parallely. This might lead to undesirable behavior.

I see the same pattern in `*TaskInfoFetcher`, we can change it there too.

Please consider retaining the
```Java
workerClient.getResults(token, MAX_RESPONSE_SIZE).get();
```
that blocks this thread until http call is complete ?

Reply:
That's a good point. Thanks for catching. Fixed.